core/src/main/java/org/jdbi/v3/core/mapper/MapMapper.java

```diff
@@ -52,32 +65,106 @@ public MapMapper(boolean foldCase) {
 
     @Override
     public RowMapper<Map<String, Object>> specialize(ResultSet rs, StatementContext ctx) throws SQLException {
-        ResultSetMetaData m = rs.getMetaData();
-        int columnCount = m.getColumnCount();
-        String[] columnNames = new String[columnCount + 1];
+        Config config = ctx.getConfig(Config.class);
+        List<String> columnNames = getColumnNames(rs, !config.getForceNewApi() && toLowerCase ? Config.CaseChange.LOWER : config.getCaseChange());
 
-        for (int i = 1; i <= columnCount; i++) {
-            String key = m.getColumnName(i);
-            String alias = m.getColumnLabel(i);
+        return (r, c) -> {
+            Map<String, Object> row = new LinkedHashMap<>(columnNames.size());
 
-            if (alias == null) {
-                alias = key;
+            for (int i = 0; i < columnNames.size(); i++) {
+                row.put(columnNames.get(i), rs.getObject(i + 1));
             }
-            if (foldCase) {
-                alias = alias.toLowerCase(Locale.ROOT);
+
+            return row;
+        };
+    }
+
+    private static List<String> getColumnNames(ResultSet rs, Config.CaseChange caseChange) throws SQLException {
+        // important: ordered and unique
+        Set<String> columnNames = new LinkedHashSet<>();
+        ResultSetMetaData meta = rs.getMetaData();
+        int columnCount = meta.getColumnCount();
+
+        for (int i = 0; i < columnCount; i++) {
+            String columnName = meta.getColumnName(i + 1);
+            String alias = meta.getColumnLabel(i + 1);
+
+            String name = caseChange.apply(alias == null ? columnName : alias);
+
+            boolean added = columnNames.add(name);
+            if (!added) {
+                throw new RuntimeException("column " + name + " appeared twice in this resultset!");
             }
+        }
 
-            columnNames[i] = alias;
+        return new ArrayList<>(columnNames);
+    }
+
+    @Beta
+    public static class Config implements JdbiConfig<Config> {
```

Comment:
Our naming convention for configuration classes is to pluralize the configured API: e.g. `SqlStatements`, `Arguments`, `RowMappers`, etc.

This convention was chosen to make inline configuration commands read like sentences:

```java
Map<String, Foo> foosById = handle.createQuery("select * from foo")
    .configure(MapEntryMappers.class, config -> config.setKeyColumn("id"))
    .collectInto(new GenericType<Map<String, Foo>>() {});
```
Let's make this a top-level class named `MapMappers`.

Reply:
done